tests/serverpod_test_server/test/generated_methods/from_json_test.dart

```diff
@@ -0,0 +1,303 @@
+import 'package:serverpod/serverpod.dart';
+import 'package:serverpod_test_server/src/generated/protocol.dart';
+import 'package:test/test.dart';
+
+void main() {
+  test(
+      'Given a non-nullable field, when deserializing from JSON with null value, then a TypeError is thrown',
+      () {
+    expect(
+      () => SimpleData.fromJson({"num": null}),
+      throwsA(isA<TypeError>()),
+    );
+  });
+
+  test(
+      'Given a class with non-nullable fields, when deserializing from an empty JSON, then a TypeError is thrown',
+      () {
+    expect(
+      () => SimpleData.fromJson({}),
+      throwsA(isA<TypeError>()),
+    );
+  });
+
+  test(
+      'Given a class with a nullable integer field, when deserializing from JSON with correct, incorrect, and null values, then proper behavior is observed',
+      () {
+    expect(
+      Types.fromJson({'anInt': 1}).anInt,
+      1,
+    );
+
+    expect(
+      () => Types.fromJson({'anInt': '1'}).anInt,
+      throwsA(isA<TypeError>()),
+    );
+
+    expect(
+      Types.fromJson({'anInt': null}).anInt,
+      null,
+    );
+  });
+
+  test(
+      'Given a class with a nullable double field, when deserializing from JSON with correct, incorrect, and null values, then it behaves as expected',
+      () {
+    expect(
+      Types.fromJson({'aDouble': 20.20}).aDouble,
+      20.20,
+    );
+
+    expect(
+      () => Types.fromJson({'aDouble': '20.20'}).aDouble,
+      throwsA(isA<TypeError>()),
+    );
+
+    expect(
+      Types.fromJson({'aDouble': null}).aDouble,
+      null,
+    );
+  });
+
+  test(
+      'Given a class with a nullable boolean field, when deserializing from JSON with correct, incorrect, and null values, then it behaves as expected',
+      () {
+    expect(
+      Types.fromJson({'aBool': true}).aBool,
+      true,
+    );
+
+    expect(
+      () => Types.fromJson({'aBool': 'true'}).aBool,
+      throwsA(isA<TypeError>()),
+    );
+
+    expect(
+      Types.fromJson({'aBool': null}).aBool,
+      null,
+    );
+  });
+
+  test(
+      'Given non-nullable DateTime, when deserializing with correct value in JSON, then the value is deserialized correctly',
+      () {
+    expect(
+      Types.fromJson({'aDateTime': '2024-01-01T00:00:00.000Z'}).aDateTime,
+      DateTime.tryParse('2024-01-01T00:00:00.000Z'),
+    );
+
+    expect(
+      Types.fromJson(
+        {'aDateTime': DateTime.tryParse('2024-01-01T00:00:00.000Z')},
+      ).aDateTime,
+      DateTime.tryParse('2024-01-01T00:00:00.000Z'),
+    );
```

Comment:
Suggestion: Here we actually have different scenarios that are not really explained by the title (one parses a correct string and one parses a datetime object).